In [44]:
# Compare Algorithms
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn import metrics
import operator


### With all features except Name

In [85]:
df = pd.read_csv("titanic.csv")

df.head()

Survived  Pclass                                               Name  \
0         0       3                             Mr. Owen Harris Braund   
1         1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2         1       3                              Miss. Laina Heikkinen   
3         1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4         0       3                            Mr. William Henry Allen   

      Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0    male  22.0                        1                        0   7.2500  
1  female  38.0                        1                        0  71.2833  
2  female  26.0                        0                        0   7.9250  
3  female  35.0                        1                        0  53.1000  
4    male  35.0                        0                        0   8.0500

In [86]:
df['Sex'] = df['Sex'].replace(['male','female'],[0,1])

df = df.drop(['Name'], axis=1)

X = df.drop(['Survived'], axis=1)
y = df['Survived']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=123)

#### LogisticRegression

In [87]:
logisticRegr = LogisticRegression()
logisticRegr.fit(X_train, y_train)

score = logisticRegr.score(X_test, y_test)

print(score)

0.8202247191011236


#### Knn

In [88]:
#test k from 1 to 50
k_range = range(1,50)

score = {}



for k in k_range:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    y_pred = knn.predict(X_test)
    score[k] = metrics.accuracy_score(y_test,y_pred)

    
all_values = score.values()
print(max(all_values))

0.7153558052434457


#### SVM

In [89]:
#Create a svm Classifier
clf = SVC(kernel='linear') # Linear Kernel

#Train the model using the training sets
clf.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test)

metrics.accuracy_score(y_test, y_pred)

0.8202247191011236

#### GaussianNB

In [90]:
gnb = GaussianNB()

#Train the model using the training sets
gnb.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = gnb.predict(X_test)

metrics.accuracy_score(y_test, y_pred)

0.8127340823970037

#### Decision Tree

In [91]:
tree = DecisionTreeClassifier()
tree.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = gnb.predict(X_test)

metrics.accuracy_score(y_test, y_pred)

0.8127340823970037

### With only Pclass and Sex

In [96]:
df = pd.read_csv("titanic.csv")



df['Sex'] = df['Sex'].replace(['male','female'],[0,1])

df = df.drop(['Name','Age','Siblings/Spouses Aboard','Parents/Children Aboard', 'Fare'], axis=1)

X = df.drop(['Survived'], axis=1)
y = df['Survived']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=123)

#### LogisticRegression

In [98]:
logisticRegr = LogisticRegression()
logisticRegr.fit(X_train, y_train)

score = logisticRegr.score(X_test, y_test)

print(score)

0.8239700374531835


#### knn

In [99]:
k_range = range(1,50)

score = {}



for k in k_range:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    y_pred = knn.predict(X_test)
    score[k] = metrics.accuracy_score(y_test,y_pred)

    
all_values = score.values()
print(max(all_values))

0.8239700374531835


#### SVM

In [100]:
#Create a svm Classifier
clf = SVC(kernel='linear') # Linear Kernel

#Train the model using the training sets
clf.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test)

metrics.accuracy_score(y_test, y_pred)

0.8239700374531835

#### GaussianNB

In [101]:
gnb = GaussianNB()

#Train the model using the training sets
gnb.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = gnb.predict(X_test)

metrics.accuracy_score(y_test, y_pred)

0.8239700374531835

#### Decision Tree

In [102]:
tree = DecisionTreeClassifier()
tree.fit(X_train, y_train)

#Predict the response for test dataset
y_pred = gnb.predict(X_test)

metrics.accuracy_score(y_test, y_pred)

0.8239700374531835

### Summary

I have tried 5 different classification models: LogisticRegression, KNN, SVM, GaussianNB, and Decision Tree with all the features for X and only  2 features with Sex & Pclass. 

Overall, with all the x features, LogisticRegression and SVM with linear kernel perform the best with an accurate rate of around 82%. When using only 2 features( Sex & Pclass), all the models returns with the same accuracy rate of 82.4%

The result matches the research paper from https://arxiv.org/ftp/arxiv/papers/1810/1810.09851.pdf which states that Sex is the most important feature for survival 